### ISSUE TITLE
Misleading error message for list assignment

### ISSUE THREAD

**Issue by GugelRobin** [microsoft/pyright#5525]
**Describe the bug**
When assignin a wrong type in `list.__setitem__` the error message complains that `__key` should be a slice and `__value` should be iterable.

**To Reproduce**
See example code.

**Expected behavior**
I similar error as you'd get for a dict.

**Code or Screenshots**
```python
x: list[type] = []
x[3] = 3
# error: Argument of type "Literal[3]" cannot be assigned to parameter "__key" of type "slice" in function "__setitem__"
#    "Literal[3]" is incompatible with "slice" (reportGeneralTypeIssues)
# error: Argument of type "Literal[3]" cannot be assigned to parameter "__value" of type "Iterable[type]" in function "__setitem__"
#    "Literal[3]" is incompatible with protocol "Iterable[type]"
#      "__iter__" is not present (reportGeneralTypeIssues)

y: dict[str, type] = {}
y["foo"] = 3 # would expect a similar error as for dict
# error: Argument of type "Literal[3]" cannot be assigned to parameter "__value" of type "type" in function "__setitem__"
#     "Literal[3]" is incompatible with "type" (reportGeneralTypeIssues)
```

**VS Code extension or command-line**
Pyright 1.1.317 & pylance v2023.7.10

**Comment by erictraut** [comment]
I explored your suggested heuristic. I found a variant of it that seems to work well. Rather than reporting the error for the overload "hast the least errors", I report the error for the overload that matches the most parameters from left to right before an error occurs. This heuristic seems to produce better results — including the error that prompted you to file this bug report in the first place. This change will be in the next release of pyright.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.319, which I just published. It will also be included in a future release of pylance.
